plugins/org.jkiss.dbeaver.net.ssh/src/org/jkiss/dbeaver/model/net/ssh/SSHUtils.java

```diff
@@ -78,4 +90,94 @@ public static boolean isKeyEncrypted(byte[] privKeyValue) {
+                } else {
+                    return null;
+                }
+            }
+        } else {
+            // seems preference path not set at all, so try to resolve it and preserve
+            return resolveDefaultKnownSshHostsFile(forceFileObjectOnFail, true);
+        }
+    }
+
+    private static File resolveDefaultKnownSshHostsFile(boolean forceFileObjectOnFail, boolean updatePreferences) {
+        try {
+            String userHomePathString = System.getProperty(IConstants.SYSTEM_PROPERTY_USER_HOME);
+            if (userHomePathString != null) {
+                Path userHomeDirPath = Paths.get(userHomePathString);
+                if (userHomeDirPath.toFile().isDirectory()) {
+                    Path sshHomeDirPath = userHomeDirPath.resolve(DEFAULT_SSH_HOME_DIR_NAME);
+                    File sshHomeDir = sshHomeDirPath.toFile();
+
+                    if (RuntimeUtils.isWindows() && (!sshHomeDir.isDirectory() || !sshHomeDir.exists())) {
+                        Path sshHomeOldDirPath = userHomeDirPath.resolve(DEFAULT_SSH_HOME_DIR_NAME_WIN_OLD);
+                        File sshHomeOldDir = sshHomeDirPath.toFile();
+                        if (sshHomeOldDir.isDirectory()) {
+                            sshHomeDirPath = sshHomeOldDirPath;
+                            sshHomeDir = sshHomeOldDir;
+                        }
+                    }
+
+                    if (sshHomeDir.isDirectory() || !sshHomeDir.exists()) {
+                        // don't need to create it until we'll need to write known hosts file
+
+                        if (updatePreferences) {
+                            Platform.getPreferencesService()
+                                    .getRootNode().node(PLATFORM_SSH_PREFERENCES_NODE)
+                                    .put(PLATFORM_SSH_PREFERENCES_SSH2HOME_KEY, sshHomeDir.getAbsolutePath());
+                        }
+
+                        Path knownSshHostsFilePath = sshHomeDirPath.resolve(KNOWN_SSH_HOSTS_FILE_NAME);
+                        return knownSshHostsFilePath.toFile();
+                    } else {
+                        log.warn("Failed to resolve default SSH known hosts file location due to invalid SSH home directory " + sshHomeDirPath.toAbsolutePath());
+                    }
+                } else {
+                    log.warn("Failed to resolve default SSH known hosts file location due to missing user home directory " + userHomeDirPath.toAbsolutePath());
+                }
+            } else {
+                log.warn("Failed to resolve default SSH known hosts file location due to missing user home system property.");
+            }
+        } catch (Throwable e) {
+            log.warn("Failed to resolve default SSH known hosts file location.", e);
+        }
+
+        if (forceFileObjectOnFail) {
+            Path forcedUserProfilePath = Paths.get(RuntimeUtils.isWindows() ? "%USERPROFILE%" : "~"); // let's pretend it'll resolve itself
```

Comment:
Please use `System.getProperty(StandardConstants#ENV_USER_HOME)` instead of this dirty hack.

Reply:
The point here is not to get null and path any string to library to let it resolve this situation itself in case of failure. For example, if user has no user home we don't want to have NPE, we want to get some error with meaningfull message from ssh library.

Reply:
I'm not so sure about it, but okay, I'll believe you.

Reply:
Well, even if it's not possible not to have user directory at all, user can accidentally remove user home system property. Then we will fail on System.getProperty(StandardConstants.ENV_USER_HOME). It will return null, we will get NPE instead of error message from the library, which we want to reach.

Reply:
It's a very exceptional situation. Let's not consider it.

Reply:
I disagree that we should not consider it, because it can happened.